TESTING.md

```diff
@@ -116,6 +117,148 @@ The `integration/happy/minimal` test is always a good start if you just want to
+
+### Creating new functional test
+
+To create new test, run following command
+
+```
+pnpm new-test
+```
+
+You'll then be asked for the name of your test and list of providers you want to run this test on. If you opt out of testing any of the providers, you'll also have to specify the reason. New test will be created under `test/functional/<name of the test>` directory.
+
+### Structure of the functional test
+
+Test consists of the 3 files:
+
+- test matrix `_matrix.ts`
+- schema template `prisma/_schema.ts`
+- test suite `tests.ts`
+
+#### Test matrix
+
+`_matrix.ts` file defines parameters for generating test suites. It can have as many parameters as necessary, but at minimum, it should define at least one provider.
+
+This matrix defines 2 test suites: one for SQLite provider and one for MongoDB, while also providing a corresponding ID string for each provider:
+
+```ts
+import { defineMatrix } from '../_utils/defineMatrix'
+
+export default defineMatrix(() => [
+  [
+    {
+      provider: 'sqlite',
+      id: 'Int @id @default(autoincrement())',
+    },
+    {
+      provider: 'mongodb',
+      id: 'String @id @default(auto()) @map("_id") @db.ObjectId',
+    },
+  ],
+])
+```
+
+If matrix has multiple dimensions, test suites will be generated for all permutations of the parameters. For example. following matrix:
+
+```ts
+import { defineMatrix } from '../_utils/defineMatrix'
+
+export default defineMatrix(() => [
+  [
+    {
+      provider: 'sqlite',
+    },
+    {
+      provider: 'postgresql',
+    },
+  ],
+  [
+    {
+      providerFeatures: '',
+    },
+
+    {
+      providerFeatures: 'improvedQueryRaw',
+    },
+  ],
+])
+```
+
+Will generate following test suites:
+
+- `{ provider: 'sqlite', providerFeatures: '' }`
+- `{ provider: 'sqlite', providerFeatures: 'improvedQueryRaw' }`
+- `{ provider: 'postgresql', providerFeatures: '' }`
+- `{ provider: 'postgresql', providerFeatures: 'improvedQueryRaw' }`
+
+#### Schema template
+
+`prisma/_schema.ts` will be used for generating an actual schema for the test suite:
+
+```ts
+import { idForProvider } from '../../_utils/idForProvider'
+import testMatrix from '../_matrix'
+
+export default testMatrix.setupSchema(({ provider }) => {
+  return /* Prisma */ `
+    generator client {
+      provider = "prisma-client-js"
+    }
+    
+    datasource db {
+      provider = "${provider}"
+      url      = env("DATABASE_URI_${provider}")
+    }
+    
+    model User {
+      id ${idForProvider(provider)}
+    }
+  `
+})
+```
+
+`setupSchema` callback receives all parameters from the matrix for a particular test suite as an argument.
+`idForProvider` is a helper function which returns a correct primary key definition for each of the supported providers.
```

Comment:
so it requires both `provider` and `id` to be present in the matrix as in the example above, right?
so `id` is kind of a magic name up there with meaning, and not just a random string?

are there others than `id`?

Reply:
No, it actually does not need the `id` in the matrix, it's just a bunch of "if provider is this, that return that hardcoded string" statements under the hood. I've added `id` to the matrix example  to demonstrate that you can have more than just `provider` there, but I see now how these two examples together make a wrong impression. I'll come up with something else for the matrix definition part.

Reply:
Good I actually asked this question then 👍

Reply:
Ahhhh, so the manual solution got replaced by a hardcoded utility. Makes so much sense now. Thanks for connecting these dots for me.